pkg/git/poll/handler.go

```diff
@@ -24,70 +32,96 @@ type Watcher interface {
 
 // Handler handles all the watches for the git repositories. These watches are pulling the latest commit every syncPeriod.
 type Handler struct {
-	client      client.Client
-	watches     map[string]Watcher
-	createWatch func(gitRepo v1alpha1.GitRepo, client client.Client) Watcher // this func creates a watch. It's a struct field, so it can be replaced for a mock in unit tests.
-	log         logr.Logger
+	client    client.Client
+	log       logr.Logger
+	scheduler quartz.Scheduler
+	fetcher   GitFetcher
 }
 
-func NewHandler(client client.Client) *Handler {
+func NewHandler(ctx context.Context, client client.Client) *Handler {
+	scheduler := quartz.NewStdScheduler()
+	scheduler.Start(ctx)
 	return &Handler{
-		client:      client,
-		watches:     make(map[string]Watcher),
-		createWatch: NewWatch,
-		log:         ctrl.Log.WithName("git-latest-commit-poll-handler"),
+		client:    client,
+		log:       ctrl.Log.WithName("git-latest-commit-poll-handler"),
+		scheduler: scheduler,
+		fetcher:   &git.Fetch{},
 	}
 }
 
-// AddOrModifyGitRepoWatch adds a new watch for the gitrepo if no watch was already present.
-// It updates the existing watch for this gitrepo if present.
-func (h *Handler) AddOrModifyGitRepoWatch(ctx context.Context, gitRepo v1alpha1.GitRepo) {
-	key := getKey(gitRepo)
-	watch, found := h.watches[key]
-	if !found {
-		h.watches[key] = h.createWatch(gitRepo, h.client)
-		h.watches[key].StartBackgroundSync(ctx)
+// AddOrModifyGitRepoPollJob adds a new scheduled job for the gitrepo if no job was already present.
+// It updates the existing job for this gitrepo if present.
+func (h *Handler) AddOrModifyGitRepoPollJob(ctx context.Context, gitRepo v1alpha1.GitRepo) {
+	gitRepoPollKey := GitRepoPollKey(gitRepo)
+	scheduledJob, err := quartzretry.GetScheduledJobWithRetries(h.scheduler, gitRepoPollKey)
+	if err != nil {
+		// job was not found
+		if gitRepo.Spec.DisableGitPolling {
+			// nothing to do if disableGitPolling is set
+			return
+		}
+		h.scheduleJob(ctx, gitRepoPollKey, gitRepo, true)
 	} else {
-		oldSyncInterval := watch.GetSyncInterval()
-		watch.UpdateGitRepo(gitRepo)
-
-		gitRepoSyncInterval := 0.0
-		if pi := gitRepo.Spec.PollingInterval; pi != nil {
-			gitRepoSyncInterval = pi.Seconds()
+		if gitRepo.Spec.DisableGitPolling {
+			// if polling is disabled, just delete the job from the scheduler
+			err = quartzretry.DeleteScheduledJobWithRetries(h.scheduler, gitRepoPollKey)
+			if err != nil {
+				h.log.Error(err, fmt.Sprintf("Error deleting job %s: %v", gitRepoPollKey.String(), err))
+			}
+			return
 		}
-
-		if oldSyncInterval != gitRepoSyncInterval {
-			watch.Restart(ctx)
+		job := scheduledJob.JobDetail().Job()
+		gitRepoPollJob, ok := job.(*GitRepoPollJob)
+		if !ok {
+			h.log.Error(fmt.Errorf("invalid job"),
+				fmt.Sprintf("Error getting Gitrepo poll job, the scheduled job: %s is not a GitRepoPollJob", job.Description()))
```

Comment:
If you implement `.String()` (see earlier comment) you can just pass the `job` here?

Reply:
`job` is `quartz.Job` and `String()` would be undefined, right?
I guess it would be better that a `quartz.Job` had `String()` instead of `Description()` but...

Reply:
This looks strange to me. Just a nit, but isn't `errors.New` preferred when we don't format the error?

More importantly, it looks like the package is using the [structured logger](https://github.com/uber-go/zap?tab=readme-ov-file#quick-start), so I'd suggest something like this for all values:

```suggestion
				"Error getting Gitrepo poll job, the scheduled job is not a GitRepoPollJob", "job", job.Description()))
```